java/arcs/sdk/android/storage/ServiceStore.kt

```diff
@@ -110,22 +112,31 @@ class ServiceStore<Data : CrdtData, Op : CrdtOperation, ConsumerData>(
     }
 
     override fun on(callback: ProxyCallback<Data, Op, ConsumerData>): Int {
+        val completion = FutureRegistrationCallback()
+
         val service = checkNotNull(storageService)
-        return service.registerCallback(object : IStorageServiceCallback.Stub() {
-            override fun onProxyMessage(proxyMessage: ByteArray) {
-                scope.launch {
-                    @Suppress("UNCHECKED_CAST")
-                    callback.invoke(
-                        proxyMessage.decodeProxyMessage() as ProxyMessage<Data, Op, ConsumerData>
-                    )
+        service.registerCallback(
+            object : IStorageServiceCallback.Stub() {
+                override fun onProxyMessage(proxyMessage: ByteArray) {
+                    scope.launch {
+                        @Suppress("UNCHECKED_CAST")
+                        callback.invoke(
+                            proxyMessage.decodeProxyMessage()
+                                as ProxyMessage<Data, Op, ConsumerData>
+                        )
+                    }
                 }
-            }
-        })
+            },
+            completion
+        )
+        return completion.get()
```

Comment:
What would happen if the service end died?

Feel there is a potential deadlock, i.e. thread pool size 2

timeline
```
thread #1: registerCallback -> scope.launch {} -> completion.get() // thread #1 is blocked
thread #2:               registerCallback -> scope.launch {} -> completion.get() // thread #2 is blocked
```
scope.launch {} for both threads can not be dispatched since both threads are blocked on waiting for onSuccess/onFailure in their scope.launch {} block. (deadlock)

Reply:
Good catch on the death issue, I think I have it fixed now.

Reply:
This potential deadlock exists in the current version as well, right?

Reply:
The answer is yes! Today I've noticed that some tests started failing due to this exact deadlock (I'm not sure why the have started failing, and they are failing only in TAP, not on the GCB runs or on my machine). 

It seems to me that this is always going to be an issue: if we want `registerCallback` to be a synchronous/blocking function that returns a value, then we will be blocking a thread to wait for the result. 

Is there a solution to this other than ensuring that the threads that drive the StorageService are not the same as the threads that drive the ArcHosts?

We could make these methods (on/off/registerCallback/unregisterCallback). This might actually be a good idea, but will require some refactoring of StorageProxy, which currently indirectly triggers a call to `on` at construction time, so this method can not be suspending with the current structure.

Reply:
Sounds good to make on/off at ServiceStore suspending.

I remember there is also a todo to make the registerCallback one-way with callback.
This would also work with some refactoring i.e. make id getter Deferred<Int>, etc.

Another idea is a global unique id generator.
So we can still use scope.launch for service, in the meantime return an unique id back.
Clients get ids back asap and the service `associates` the assigned ids with callbacks simultaneously.

Reply:
This PR is the one making the registerCallback oneway w/ callback :-)

Reply:
I have some work in flight to convert `on` to a suspending method, then this PR can be converted to use Deferred instead of future.

Reply:
Cool! So glad we caught this and making it right :-)

Reply:
These are now both suspending.